src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/EditorAsyncCompletionItemManager.cs

```diff
@@ -0,0 +1,614 @@
+            string filterText,
+            ImmutableArray<EditorCompletion.CompletionFilterWithState> filters,
+            EditorCompletion.CompletionTriggerReason triggerReason,
+            List<FilterResult> itemsInList,
+            bool mustSetSelection)
+        {
+            var completionService = document.GetLanguageService<CompletionService>();
+            if (completionService == null)
+            {
+                var listWithSelections = GetHighlightedList(itemsInList, filterText);
+                return new EditorCompletion.FilteredCompletionModel(listWithSelections.ToImmutableArray(), 0, filters, EditorCompletion.UpdateSelectionHint.NoChange, centerSelection: true, uniqueItem: default);
+            }
+
+            var matchingItems = itemsInList
+                .Where(r => r.MatchedFilterText)
+                .Select(t => GetOrCreateRoslynItem(t.CompletionItem))
+                .AsImmutable();
+
+            var chosenItems = completionService.FilterItems(document, matchingItems, filterText);
+
+            var recentItems = _recentItems;
+            var bestItem = GetBestItemBasedOnMRU(chosenItems, recentItems);
+            var highlightedList = GetHighlightedList(itemsInList, filterText).ToImmutableArray();
+            var updatedFilters = GetUpdatedFilters(sortedList, itemsInList, filters, filterText);
+            EditorCompletion.CompletionItem uniqueItem = null;
+            int selectedItemIndex = 0;
+
+            // TODO: Can we get away with less complexity here by only doing hard select on preselection and not on regular filter text matching / etc...
+            // https://github.com/dotnet/roslyn/issues/29108
+
+            if (bestItem != null)
+            {
+                selectedItemIndex = itemsInList.IndexOf(i => Equals(GetOrCreateRoslynItem(i.CompletionItem), bestItem));
+                if (selectedItemIndex > -1 && bestItem != null && matchingItems.Length == 1 && filterText.Length > 0)
+                {
+                    uniqueItem = highlightedList[selectedItemIndex].CompletionItem;
+                }
+            }
+
+            var updateSelectionHint = GetUpdateSelectionHint(bestItem, snapshot, caretPosition, triggerReason, filterText, selectedItemIndex, mustSetSelection);
+
+            if (selectedItemIndex == -1)
+            {
+                selectedItemIndex = 0;
+            }
+
+            return new EditorCompletion.FilteredCompletionModel(highlightedList, selectedItemIndex, updatedFilters, updateSelectionHint, centerSelection: true, uniqueItem);
+        }
+
+        private static RoslynCompletionItem GetOrCreateRoslynItem(EditorCompletion.CompletionItem item)
+        {
+            if (item.Properties.TryGetProperty<RoslynCompletionItem>(CompletionItemSource.RoslynItem, out var roslynItem))
+            {
+                return roslynItem;
+            }
+
+            return RoslynCompletionItem.Create(item.DisplayText, item.FilterText, item.SortText);
```

Comment:
why is this item not already in hte item properties?  and, if this is the place where we need to make it, why not add it into the properties bag?

I really don't get the lifetimes between VS completion items and roslyn items.

Reply:
We can be called in the middle of the process. For example, the debugger may provide all items for further processing. In this case, those items do not have Roslyn items inside. We must consider each call to be made as a first call to Roslyn.

So, it is great if we can find anything we already pre-calculated on previous steps. However, we cannot guarantee that this always happens.

---
In reply to: [222088040](https://github.com/dotnet/roslyn/pull/29016#discussion_r222088040) [](ancestors = 222088040)